22/2.rs

```diff
@@ -0,0 +1,276 @@
+        )
+        .map(|((name, expected), result)| 
+            (
+                *name,
+                result
+                    .and_then(|actual| if *expected == actual {
+                        Ok(())
+                    } else {
+                        error(format!("Expected {} but got {}", expected, actual))
+                    })
+            )
+        )
+        .collect::<Vec<(&str, AoC<()>)>>();
+    results.iter()
+        .for_each(|(name, result)| match result {
+            Ok(()) => println!("Example {} passed.", name),
+            Err(msg) => println!("Example {} failed: {}.", name, msg)
+        });
+
+    if results.iter().any(|t| t.1.is_err()) {
+        panic!("Please address errors before attempting the problem.")
+    }
+
+    println!(
+        "{}",
+        operation(format!("{}/input", DAY)).expect("Unexpected Error in main input.")
+    );
+}
+
+fn errorize<S: Into<Box<dyn Error>>>(msg: S) -> Box<dyn Error> {
+    return msg.into();
+}
+
+fn error<T, S: Into<Box<dyn Error>>>(err: S) -> AoC<T> {
+    return Err(err.into());
+}
+
+fn operation(filename: String) -> AoC<Solution> {
+    return fs::read_to_string(filename)
+        .map_err(errorize)
+        .and_then(parse)
+        .and_then(solve);
+}
+
+#[derive(PartialEq, Eq)]
+struct Command {
+    on: bool,
+    points: Points
+}
+
+impl Command {
+    fn from_input(line: &str) -> AoC<Command> {
+        let mut bits = line.split(&[' ', ','][..]);
+        let on = bits.next()
+            .ok_or(errorize("No on/off"))
+            .and_then(|on| match on {
+                "on" => Ok(true),
+                "off" => Ok(false),
+                other => error(format!("Unrecognized on/off: '{}'", other))
+            })?;
+        let range_strs: Vec<&str> = bits.take(3).collect();
+        if range_strs.len() != 3 {
+            return error(format!("Unexpected number of ranges: {}", range_strs.len()));
+        }
+
+        let ranges = range_strs.iter().zip(["x=", "y=", "z="])
+            .map(|(string, prefix)| parse_range(prefix, string))
+            .collect::<AoC<Vec<Range>>>()?;
+        return match &ranges[..] {
+            [x, y, z] => Ok(Command {
+                on: on,
+                points: Points {
+                    xs: x.clone(),
+                    ys: y.clone(),
+                    zs: z.clone()
+                }
+            }),
+            _ => error("I don't know how this happened"),
+        }
+    }
+}
+
+fn parse_range(prefix: &str, line: &str) -> AoC<Range> {
+    let bounds = line.strip_prefix(prefix)
+        .ok_or(errorize(format!("Expected prefix '{}' on '{}'", prefix, line)))
+        .and_then(|rest| rest.split("..")
+            .map(str::parse)
+            .collect::<Result<Vec<isize>, _>>()
+            .map_err(errorize)
+        )?;
+
+    if bounds.len() == 2 {
+        return Ok(Range::new(bounds[0], bounds[1]))
+    }
+
+    return error(format!("Got the wrong number of bounds from '{}'.", line));
```

Comment:
Gently more idiomatic here to use a `match`:
```
match bounds.len() {
  2 => Ok(Range::new(bounds[0], bounds[1])),
  wrong => error(format!("Got the wrong number of bounds from '{}': {}", line, wrong)),
}
```
But also, you can destructure slices! 
```
// Take a read-only slice of the entire vec
match &bounds[..] {
  [bound_start, bound_stop] => Ok(Range::new(bound_start, bound_stop)),
  wrong => // etc
}
```